### PR TITLE
:recycle: Improved jsonschema exporter

### PR DESCRIPTION
**Pull request by sschleemilch**
# Improved the jsonschema exporter

- Added typed arrays using `items`

  ```yaml
  Vehicle:
    type: branch
    description: Vehicle
  
  Vehicle.Attribs:
    type: attribute
    description: Attribs
    datatype: string[]
  ```
  
  Before:
  ```json
  {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "Vehicle",
    "type": "object",
    "description": "Vehicle",
    "additionalProperties": false,
    "properties": {
      "Attribs": {
        "description": "Attribs",
        "type": "array"
      }
    }
  }
  ```
  Now:
  ```json
  {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "Vehicle",
    "type": "object",
    "description": "Vehicle",
    "properties": {
      "Attribs": {
        "type": "array",
        "description": "Attribs",
        "items": {
          "type": "string"
        }
      }
    },
  }
  ```
- Added support for custom types a.k.a `structs`:
  ```yaml
  Vehicle:
    type: branch
    description: Vehicle
  
  Vehicle.Doors:
    type: attribute
    description: Doors
    datatype: Types.Door[]
  ```

  ```yaml
  Types:
    type: branch
    description: Types
  
  Types.Door:
    type: struct
    description: Door
  
  Types.Door.Windows:
    type: property
    description: Windows
    datatype: uint8
  ```

  ```json
  {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "Vehicle",
    "type": "object",
    "description": "Vehicle",
    "properties": {
      "Doors": {
        "type": "array",
        "description": "Doors",
        "items": {
          "type": "object",
          "description": "Door",
          "properties": {
            "Windows": {
              "type": "integer",
              "description": "Windows",
              "minimum": 0,
              "maximum": 255
            }
          },
          "additionalProperties": false
        }
      }
    },
    "additionalProperties": false
  }
  ```

- Including `minimum` and `maximum` automatically for types to express their max mins (only until uint/int64 exclusive)
  ```yaml
  Vehicle:
    type: branch
    description: Vehicle
  
  Vehicle.Speed:
    type: sensor
    description: Speed
    datatype: int16
  ```
  ```json
  {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "Vehicle",
    "type": "object",
    "description": "Vehicle",
    "properties": {
      "Speed": {
        "type": "integer",
        "description": "Speed",
        "minimum": -32768,
        "maximum": 32767
      }
    },
    "additionalProperties": false
  }
  ```
- Fixed support for `numeric`


# Added

- Model check that `min` and `max` are in the range of the datatype
  ```yaml
  Vehicle:
    type: branch
    description: Vehicle
  
  Vehicle.Something:
    type: sensor
    description: Something
    datatype: int8
    max: 300
  ```
  ```bash
    ❯ vspec export tree -s test.vspec
  [10:26:23] WARNING  No 'quantity' files defined. Default not existing: /Users/Q534774/workspace/vss-tools/quantities.yaml                                                                                                                             main.py:60
             WARNING  No 'unit' files defined. Default not existing: /Users/Q534774/workspace/vss-tools/units.yaml                                                                                                                                      main.py:66
             INFO     VSpecs loaded, amount=1                                                                                                                                                                                                         vspec.py:130
             CRITICAL 'Vehicle.Something' has 1 model error(s):                                                                                                                                                                                        main.py:234
                      [                                                                                                                                                                                                                                           
                          {                                                                                                                                                                                                                                       
                              'type': 'assertion_error',                                                                                                                                                                                                          
                              'loc': (),                                                                                                                                                                                                                          
                              'msg': "Assertion failed, max '300' is not an 'int8'",                                                                                                                                                                              
                              'input': {                                                                                                                                                                                                                          
                                  'fqn': 'Vehicle.Something',                                                                                                                                                                                                     
                                  'type': <NodeType.SENSOR: 'sensor'>,                                                                                                                                                                                            
                                  'description': 'Something',                                                                                                                                                                                                     
                                  'comment': None,                                                                                                                                                                                                                
                                  'delete': False,                                                                                                                                                                                                                
                                  'deprecation': None,                                                                                                                                                                                                            
                                  'constUID': None,                                                                                                                                                                                                               
                                  'fka': [],                                                                                                                                                                                                                      
                                  'instantiate': True,                                                                                                                                                                                                            
                                  'datatype': 'int8',                                                                                                                                                                                                             
                                  'max': 300                                                                                                                                                                                                                      
                              }                                                                                                                                                                                                                                   
                          }                                                                                                                                                                                                                                       
                      ]                                                                                                                                                                                                                                           

  ```

### CODE REVIEW

**Review by SebastianSchildt — APPROVED** [review]
This makes sense to me, but I am also not using jsonschema currently so it does not break anything for me :)

Do we need backwards compatibility here? (I don't)

**Comment by erikbosch** [comment]
MoM:

- Erik wants to review
- Ok to merge if no remarks after the review from Erik

File: src/vss_tools/exporters/jsonschema.py [hunk 1]
```diff
@@ -189,39 +106,90 @@ def cli(
...
+    if isinstance(node.data, VSSDataDatatype):
+        ref = schema
+        if is_array(node.data.datatype):
+            schema["type"] = "array"
+            schema["items"] = {}
+            ref = schema["items"]
+        datatype = node.data.datatype.rstrip("[]")
+        if datatype in type_map:
+            target_type = type_map[datatype]
+            target_type = type_map[datatype]
```
Comment (erikbosch) [inline src/vss_tools/exporters/jsonschema.py:159]: Duplicate of previous line

File: src/vss_tools/exporters/jsonschema.py [hunk 2]
```diff
@@ -189,39 +106,90 @@ def cli(
...
+            ref["type"] = target_type[0]
+            if len(target_type) > 1:
```
Comment (erikbosch) [inline src/vss_tools/exporters/jsonschema.py:161]: As of today we expect "len" to be either 1 or 3, right? I.e. never anything else.

**Comment by sschleemilch** [comment]
> Do we need backwards compatibility here? (I don't)

Nope, old ones were not including array restrictions and the new one is. So it will definitely break

File: src/vss_tools/exporters/jsonschema.py (hunk 1 above)
Comment (sschleemilch) [reply src/vss_tools/exporters/jsonschema.py:159]: Good catch. Fixed

File: src/vss_tools/exporters/jsonschema.py (hunk 2 above)
Comment (sschleemilch) [reply src/vss_tools/exporters/jsonschema.py:161]: Right now it only will be 1 or 3. Yeah. Thought about specifying at least the lower boundary for uint64. Then we would have 2 :)

**Review by erikbosch — APPROVED** [review]
LGTM

**Pull request merged**